### PR TITLE
[Feature] Support visualization for single gpu test

### PR DESCRIPTION
**Pull request by ToumaKazusa3**

### CODE REVIEW

File: mmtrack/models/sot/base.py [hunk 1]
```diff
@@ -243,12 +243,15 @@ def show_result(self,
...
-                    out_file=None):
+                    out_file=None,
+                    **kwargs):
         """Visualize tracking results.
 
         Args:
             img (str or ndarray): The image to be displayed.
-            result (ndarray): ndarray of shape (4, ).
+            result (dict): Tracking result.
+                The value of key 'track_results' is ndarray with shape (n, 5)
```
Comment (GT9505) [inline mmtrack/models/sot/base.py:253]: the shape of sot is `(5, )`

File: mmtrack/models/sot/base.py [hunk 2]
```diff
@@ -265,11 +268,15 @@ def show_result(self,
...
             ndarray: Visualized image.
         """
-        assert result.ndim == 1
-        assert result.shape[0] == 4
+        assert isinstance(result, dict)
+        track_result = result.get('track_results', None)
+        assert track_result.ndim == 1
+        assert track_result.shape[0] == 5
+
+        track_result = track_result[:4]
```
Comment (GT9505) [inline mmtrack/models/sot/base.py:276]: `track_bbox = track_result[:4]`

File: mmtrack/models/vid/base.py [hunk 3]
```diff
@@ -251,8 +251,9 @@ def show_result(self,
 
         Args:
             img (str or Tensor): The image to be displayed.
-            result (Tensor or tuple): The results to draw over `img`
-                bbox_result or (bbox_result, segm_result).
+            result (dict): Detection result. The value of key 'bbox_results'
+                is list with length 30, and each element in list is ndarray
+                with shape(n, 5) in [tl_x, tl_y, br_x, br_y, score] format.
```
Comment (GT9505) [inline mmtrack/models/vid/base.py:256]: remeber to modify here

File: mmtrack/apis/test.py [hunk 4]
```diff
@@ -7,45 +8,98 @@
...
+        show_score_thr (float, optional): The score threshold of visualization.
+            Defaults to 0.3.
 
     Returns:
         dict[str, list]: The prediction results.
     """
     model.eval()
     results = defaultdict(list)
     dataset = data_loader.dataset
+    last_img_meta = None
```
Comment (GT9505) [inline mmtrack/apis/test.py:41]: delete this line

File: mmtrack/apis/test.py [hunk 5]
```diff
@@ -7,45 +8,98 @@
...
+                    img_dirs,
+                    f'{img_dirs}/out_video.mp4',
+                    fps=fps,
+                    fourcc='mp4v',
+                    filename_tmpl=filename_tmpl,
+                    start=start_frame_id,
+                    end=end_frame_id,
+                    show_progress=False)
+
+            last_img_meta = img_meta
```
Comment (GT9505) [inline mmtrack/apis/test.py:98]: change it to `prev_img_meta`

File: mmtrack/apis/test.py [hunk 6]
```diff
@@ -7,45 +8,98 @@
...
-        show_score_thr (float): The score threthold of visualization (Not
-            supported for now). Defaults to 0.3.
+        show (bool, optional): If True, visualize the prediction results.
+            Defaults to False.
+        out_dir (str, optional): Path of directory to save the
+            visualization results. Defaults to None.
+        fps (int, optional): FPS of the output video.
+            Defaults to 3.
+        show_score_thr (float, optional): The score threshold of visualization.
+            Defaults to 0.3.
```
Comment (GT9505) [inline mmtrack/apis/test.py:33]: add a docsting `Only used in VID for now.`

File: mmtrack/apis/test.py [hunk 7]
```diff
@@ -7,45 +8,98 @@
...
+                result,
+                show=show,
+                out_file=out_file,
+                score_thr=show_score_thr)
+
+            # generate the video from images
+            # case1: the current frame is the first frame of the new video,
+            # except the first frame of the entire dataset.
+            # case2: the current frame is the last frame of the entire dataset.
+            if i != 0 and img_meta['frame_id'] == 0 or i == len(dataset):
```
Comment (GT9505) [inline mmtrack/apis/test.py:78]: add `if out_dir`

File: mmtrack/apis/test.py [hunk 8]
```diff
@@ -7,45 +8,98 @@
...
+                img_prefix, img_name = last_img_meta['ori_filename'].rsplit(
```
Comment (GT9505) [inline mmtrack/apis/test.py:79]: add `prev_` to all variables as perfix under this `if`.
for example, `prev_img_name`

File: mmtrack/apis/test.py [hunk 9]
```diff
@@ -7,45 +8,102 @@
...
+                score_thr=show_score_thr)
+
+            # Generate a video from images.
+            if out_dir:
+                # The frame_id == 0 means the model starts processing
+                # a new video, therefore we can write the previous video.
+                # There are two corner cases.
+                # Case 1: i == 0 means there is no previous video.
+                # Case 2: i == len(dataset) means processing the last video
+                if i != 0 and img_meta['frame_id'] == 0 or i == len(dataset):
```
Comment (GT9505) [inline mmtrack/apis/test.py:81]: please don't use 2 `if`.
`need_write_video = (pre_img_meta is not None and img_meta['frame_id'] == 0 or i == len(dataset))`
`if out_dir and need_write_video:`
remeber to modify the corresponding comments

**Review by GT9505 — APPROVED** [review]

**Review by OceanPang — APPROVED** [review]

**Pull request merged**